cpp/src/arrow/dataset/scanner.h

```diff
@@ -32,6 +32,7 @@
 #include "arrow/memory_pool.h"
 #include "arrow/type_fwd.h"
 #include "arrow/util/async_generator.h"
+#include "arrow/util/thread_pool.h"
```

Comment:
No need to, `internal::Executor` is already declared in `arrow/util/type_fwd.h`.
Also, can you try to remove `async_generator.h`?

Reply:
I tried and failed removing `async_generator.h` before.  The crux of the issue was...

`AsyncGenerator<T>` is a typedef of `std::function<Future<T>()>` and not a proper class in its own right (in fact, `async_generator.h` hides all generator classes behind `MakeXYZGenerator` and so it doesn't actually need to export any types beyond `std::function`.

`std::function<Future<T>()>` requires `<functional>`.

We don't want `<functional>` in a `type_fwd`.

I'm open to suggestions.

Reply:
Perhaps we can use `std::function<Future<T>()>` directly instead of the typedef. @bkietz what do you think?

Reply:
Sorry, I was not clear with my language.  By "no types beyond `std::function`" I meant "no classes or structs".  We would still need to include `async_generator.h` for the functions (e.g. `MakeMappedGenerator`).

So without the typedef you still have the problem that you cannot forward declare the `MakeXyz` functions without including `<functional>` in the forward declaration.

Reply:
But the `MakeXyz` functions are not used in `scanner.h`, are they?

Reply:
Well, you caught me out, they're not all named `MakeXyz` :smile: 

```
      auto scan_fut = CollectAsyncGenerator(std::move(scan_gen))
                          .Then([state, id](const RecordBatchVector& rbs) {
                            state->Emplace(rbs, id);
```

This call to `CollectAsyncGenerator` requires the include.

Reply:
I don't see any call to `CollectAsyncGenerator` in `scanner.h`. Am I missing something?

Reply:
Ah, it is in `scanner.cc`.  My mistake.  Yes, I can move this include to `scanner.cc`.

Reply:
Done.